### ISSUE TITLE
Is it possible to generalize these functions?

### ISSUE THREAD

**Issue by erikd** [mitchellwrosen/text-ansi#3]
Currently things like `red` and `blue` are of type `Text -> Text` .

Is it possible to generalize these so they are of type `IsString a => a -> a`. That would make it easier to use functions in this package with things like `Text.Printf` .

**Comment by mitchellwrosen** [comment]
Sure, that sounds like a reasonable module to add alongside the others. I don't think exporting only a generalized interface would be wise - if you already have a `Text` or `Text.Builder` in hand, we wouldn't want to round-trip through `IsString` to get right back to a `Text.Builder` ;)

What should the module name be called? Right now there's:

- `Text.ANSI`
- `Text.Builder.ANSI`
- `Text.Lazy.Builder.ANSI`

**Comment by erikd** [comment]
Isn't this just a generalization? If so, can't the existing definitions just be changed?

**Comment by mitchellwrosen** [comment]
In my previous comment, I thought I had explained why we probably shouldn't just generalize the API, but when I went to elaborate I realized the justification was inaccurate, or at least poorly thought out 😅. So here's another attempt:

For efficiency, this library styles text with a strict text builder internally, and has a rewrite rule so that chains of stylings (e.g. `red . bold`) only go through the underlying builder once.

Let's just take the simple case of a single styling, which surrounds its argument with some escape sequences (say `^` and `$`):

```haskell
red :: Text -> Text
red s = build (fromString "^" <> text s <> fromString "$")
```

Were we to generalize this, I we'd have to give up on the intermediate builder.

```haskell
red :: (IsString a, Monoid a) => a -> a
red s = fromString "^" <> s <> fromString "$"
```

Certainly not the end of the world to just append texts/strings with an inefficient `<>`; it's what you'd get with other APIs like `ansi-terminal`. But it does feel unwise to me to punish users who have a concrete Text-ish type that can be converted to a strict text builder, when a new module could be provided with the overloaded versions instead.

**Comment by erikd** [comment]
For a module name, maybe `StringLike.ANSI` or `Text.ANSI.StringLike` ?

**Comment by mitchellwrosen** [comment]
Thinking about this more. Could you explain the use case one more time? I'm not sure I got the connection to `Text.Printf`.

**Comment by erikd** [comment]
I have code which does things like:
```
printf "%6.3f %s" a (Text.unpack . red . Text.pack $ printf "6.3f" b)
```
just waiting for the generalised version so I can do:
```
printf "%6.3f %s" a (red $ printf "6.3f" b)
```

**Comment by mitchellwrosen** [comment]
Ah, hm. That looks like

```haskell
red :: String -> String
```

to me, no?

**Comment by mitchellwrosen** [comment]
(Alternatively, `instance PrintfArg Text`?)

**Comment by erikd** [comment]
Is `red :: String -> String` available somewhere?

**Comment by mitchellwrosen** [comment]
No, but I think it makes sense to add to this package.

**Comment by erikd** [comment]
Would be nice to have 6 or 8 `String -> String` colors was well as an `rgb` function so the user can define their own.

**Comment by mitchellwrosen** [comment]
Great! This is now available in `text-ansi-0.2.1`:

https://hackage.haskell.org/package/text-ansi-0.2.1/docs/String-ANSI.html

**Comment by erikd** [comment]
Thanks!
